apm-agent-plugins/apm-slf4j-plugin/src/main/java/co/elastic/apm/agent/slf4j/Slf4JMdcActivationListener.java

```diff
@@ -68,37 +78,79 @@ public MethodHandle get(ClassLoader classLoader) {
+                try {
+                    return MethodHandles.lookup()
+                        .findStatic(classLoader.loadClass(ORG_SLF4J_MDC), "get", MethodType.methodType(String.class, String.class));
+                } catch (Exception ignore) {
+                    // this class loader does not have the slf4j api
+                    return null;
+                }
+            }
+        });
     @Nullable
     private LoggingConfiguration config;
+    @Nullable
+    private ElasticApmTracer tracer;
 
     @Override
     public void init(ElasticApmTracer tracer) {
+        this.tracer = tracer;
         config = tracer.getConfig(LoggingConfiguration.class);
     }
 
     @Override
     public void onActivate(TraceContextHolder<?> context) throws Throwable {
         if (config != null && config.isLogCorrelationEnabled() && context.isSampled()) {
-            MethodHandle put = mdcPutMethodHandleCache.get(Thread.currentThread().getContextClassLoader());
-            TraceContext traceContext = context.getTraceContext();
+            ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+
+            // always push something to span ID stack - either valid ID or a non-valid-id marker
+            String formerSpanId = NA_SPAN_ID;
+            MethodHandle get = mdcGetMethodHandleCache.get(contextClassLoader);
+            if (get != null) {
+                formerSpanId = (String) get.invokeExact(SPAN_ID);
+                if (formerSpanId == null) {
+                    formerSpanId = NA_SPAN_ID;
+                }
+            }
+            spanIdStack.get().push(formerSpanId);
+
+            MethodHandle put = mdcPutMethodHandleCache.get(contextClassLoader);
             if (put != null) {
-                put.invokeExact(TRACE_ID, traceContext.getTraceId().toString());
-                String idKey = context instanceof Span ? SPAN_ID : TRANSACTION_ID;
-                put.invokeExact(idKey, traceContext.getId().toString());
+                TraceContext traceContext = context.getTraceContext();
+                put.invokeExact(SPAN_ID, traceContext.getId().toString());
+                if (tracer != null && tracer.isThreadRootContext(context)) {
+                    put.invokeExact(TRACE_ID, traceContext.getTraceId().toString());
+                    put.invokeExact(TRANSACTION_ID, traceContext.getTransactionId().toString());
+                }
             }
         }
     }
 
     @Override
-    public void onDeactivate() throws Throwable {
-        if (config != null && config.isLogCorrelationEnabled()) {
-            MethodHandle remove = mdcRemoveMethodHandleCache.get(Thread.currentThread().getContextClassLoader());
+    public void onDeactivate(TraceContextHolder<?> context) throws Throwable {
+        if (config != null && config.isLogCorrelationEnabled() && context.isSampled()) {
+            String formerSpanId = spanIdStack.get().poll();
```

Comment:
Instead of maintaining another stack here, couldn't we just get `tracer.getActive()` and call `isSampled()` and `traceContext.getId().toString()` on that instance?

Reply:
This contradicts what you mentioned in the above comment :)
It may be already ended and recycled, so not allowed to observe at this point.

Reply:
What I meant was to get the parent of the context which is being deactivated. You reversed the order in the PR so that it deactivates after calling the listeners. I think we should make it more obvious from the method name if the deactivation happens before or after. For example we should call it either `beforeDeactivate` or `afterDeactivate`.

Reply:
Ahh, I see. 
To me, the right way to undo/reverse `A -> B -> C -> D` is `!D -> !C -> !B -> !A`. So, for this matter, if `A` is `TraceContextHolder#activate` and `B` is `ActivationListener#onActivate`, It seemed to make sense to reverse the order. 
What if I change `ActivationListener#onActivate` to `beforeActivate` and `ActivationListener#onDeactivate` to `afterDeactivate`? Then I can remove the additional stack as you suggest and have consistent code in both methods: 
```
    public void beforeActivate(TraceContextHolder<?> context) throws Throwable {
        ...
        if (tracer != null) {
            if (tracer.getActive() == null) {
                put.invokeExact(TRACE_ID, traceContext.getTraceId().toString());
                put.invokeExact(TRANSACTION_ID, traceContext.getTransactionId().toString());
            }
            put.invokeExact(SPAN_ID, traceContext.getId().toString());
        }
    }

    public void afterDeactivate(TraceContextHolder<?> context) throws Throwable {
        ...
        if (tracer != null) {
            TraceContextHolder active = tracer.getActive();
            if (active == null) {
                remove.invokeExact(TRACE_ID);
                remove.invokeExact(TRANSACTION_ID);
            } else if (active.isSampled()) {
                put.invokeExact(SPAN_ID, active.getTraceContext().getId().toString());
            }
        }
    }
```